reaktive/src/commonMain/kotlin/com/badoo/reaktive/single/SingleByEmitter.kt

```diff
@@ -1,35 +1,33 @@
 package com.badoo.reaktive.single
 
+import com.badoo.reaktive.base.tryCatch
 import com.badoo.reaktive.disposable.Disposable
 import com.badoo.reaktive.disposable.DisposableWrapper
 import com.badoo.reaktive.disposable.doIfNotDisposed
 
 fun <T> single(onSubscribe: (emitter: SingleEmitter<T>) -> Unit): Single<T> =
     singleUnsafe { observer ->
-        val emitter =
-            object : DisposableWrapper(), SingleEmitter<T> {
-                override fun onSuccess(value: T) {
-                    doIfNotDisposed(dispose = true) {
-                        observer.onSuccess(value)
-                    }
-                }
-
-                override fun onError(error: Throwable) {
-                    doIfNotDisposed(dispose = true) {
-                        observer.onError(error)
-                    }
-                }
-
-                override fun setDisposable(disposable: Disposable) {
-                    set(disposable)
-                }
-            }
-
+        val emitter = DisposableEmitter(observer)
         observer.onSubscribe(emitter)
+        emitter.tryCatch(block = { onSubscribe(emitter) })
+    }
 
-        try {
-            onSubscribe(emitter)
-        } catch (e: Throwable) {
-            emitter.onError(e)
+private class DisposableEmitter<T>(
```

Comment:
No, what is the point to do it? I wanted to remove copy-pasted
```kotlin
override fun setDisposable(disposable: Disposable) {
        set(disposable)
}
```
block by delegating it to common base class.

Reply:
I don't understand, please elaborate

Reply:
Check original comment. I proposed to introduce common base class for all 4 cases:
```kotlin
abstract class EmitterDisposableWrapper: DisposableWrapper(), Emitter {
    override fun setDisposable(disposable: Disposable) {
        set(disposable)
    }
}
...
val emitter = object : EmitterDisposableWrapper(), CompletableEmitter { }
val emitter = object : EmitterDisposableWrapper(), MaybeEmitter { }
...
```

Reply:
Fixed! Thanks!